bugwarrior/services/gitlab.py

```diff
@@ -70,6 +74,181 @@ def default_priorities(cls, values):
         return values
 
 
+class GitlabClient(ServiceClient):
+    """Abstraction of Gitlab API v4"""
+
+    def __init__(self, host, token, use_https, verify_ssl):
+        if use_https:
+            self.scheme = 'https'
+        else:
+            self.scheme = 'http'
+        self.verify_ssl = verify_ssl
+
+        self.host = host
+        self.token = token
+
+    def _base_url(self):
+        return f"{self.scheme}://{self.host}/api/v4/"
+
+    def _fetch(self, relative_url: str, **kwargs) -> dict:
+        """Perform a fetch operation on the gitlab server
+
+        :param relative_url: This part will be appended to the base api URL for the call
+        :type relative_url: str
+        :param kwargs: will be sent alongside the request.get call
+        :rtype: dict
+        """
+        headers = {'PRIVATE-TOKEN': self.token}
+        url = self._base_url() + relative_url
+
+        if not self.verify_ssl:
+            requests.packages.urllib3.disable_warnings()
+        response = requests.get(
+            url, headers=headers, verify=self.verify_ssl, **kwargs)
+
+        return self.json_response(response)
+
+    def _fetch_paged(self, relative_url: str, page_size: int = 100) -> list:
+        """Make a gitlab REST API call with pagination. Calls will be buffered on pages with size
+        ``page_size``.
+
+        :param relative_url:
+        :type relative_url: str
+        :param page_size: Size of fetch pages. Defaults to 100
+        :type page_size: int
+        :rtype: list
+        """
+        params = {
+            'page': 1,
+            'per_page': page_size,
+        }
+
+        full = []
+        detect_broken_gitlab_pagination = []
+        while True:
+            items = self._fetch(relative_url, params=params)
+            if not items:
+                break
+
+            # XXX: Some gitlab versions have a bug where pagination doesn't
+            # work and instead return the entire result no matter what. Detect
+            # this by seeing if the results are the same as the last time
+            # around and bail if so. Unfortunately, while it is a GitLab bug,
+            # we have to deal with instances where it exists.
+            if items == detect_broken_gitlab_pagination:
+                break
+            detect_broken_gitlab_pagination = items
+
+            full += items
+
+            if len(items) < params['per_page']:
+                break
+            params['page'] += 1
+
+        return full
+
+    def get_repos(self, include_repos=None, include_regex=None,
+                  only_membership=False, only_owned=False):
```

Comment:
Sure!

Reply:
I've changed that function in 2744b96 a bit. As the `include_regex` parameter was not really used inside the function, I decided to drop it. In the old structure it was used to only use `include_repos` if `include_regex` was not set in the config. So basically, this seems as if it was just used, because it was there. However, in my opinion the decision whether `include_repos` should be used or not belongs into the `Service`, not the `Client`.